File: internal/hemcore/tar/tar.go
package tar

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/qyinm/hem/internal/hemcore/types"
)

const (
	blockSize      = 512
	headerSize     = 512
	magic          = "ustar"
	maxEntryBytes  = 500 * 1024 * 1024
)

// Error represents tar read/write failures.
type Error struct {
	Message string
	Cause   error
}

func (e *Error) Error() string {
	if e.Cause != nil {
		return e.Message + ": " + e.Cause.Error()
	}
	return e.Message
}

// WriteTar writes a tar archive and returns the SHA-256 hex digest of the archive bytes.
func WriteTar(entries []types.TarEntry, outputPath string) (string, error) {
	var blocks []byte

	for _, entry := range entries {
		header, err := encodeHeader(entry)
		if err != nil {
			return "", err
		}
		blocks = append(blocks, header...)

		if entry.EntryType == types.TarEntryFile && len(entry.Content) > 0 {
			blocks = append(blocks, entry.Content...)
			padding := blockSize - (len(entry.Content) % blockSize)
			if padding < blockSize {
				blocks = append(blocks, make([]byte, padding)...)
			}
		}
	}

	blocks = append(blocks, make([]byte, blockSize*2)...)

	if err := os.MkdirAll(filepath.Dir(outputPath), 0o755); err != nil {
		return "", &Error{Message: "create output directory", Cause: err}
	}
	if err := os.WriteFile(outputPath, blocks, 0o644); err != nil {
		return "", &Error{Message: "write tar archive", Cause: err}
	}

	return hexDigest(blocks), nil
}

// ReadTar reads a tar archive from disk.
func ReadTar(inputPath string) ([]types.TarEntry, string, error) {
	archive, err := os.ReadFile(inputPath)
	if err != nil {
		return nil, "", &Error{Message: "read tar archive", Cause: err}
	}

	checksum := hexDigest(archive)
	var entries []types.TarEntry
	offset := 0

	for offset+headerSize <= len(archive) {
		headerBlock := archive[offset : offset+headerSize]

		if isZeroBlock(headerBlock) {
			offset += blockSize
			if offset+blockSize <= len(archive) && isZeroBlock(archive[offset:offset+blockSize]) {
				break
			}
			continue
		}

		header, err := decodeHeader(headerBlock)
		if err != nil {
			return nil, "", err
		}
		offset += headerSize

		if header.typeflag != '0' && header.typeflag != '5' {
			return nil, "", &Error{Message: fmt.Sprintf(
				`Unsupported tar entry type "%c" for "%s". Only regular files (typeflag '0') and directories (typeflag '5') are accepted.`,
				header.typeflag, header.name,
			)}
		}

		if header.linkname != "" {
			return nil, "", &Error{Message: fmt.Sprintf(
				`Tar entry "%s" has a non-empty linkname field ("%s"). Link entries are not supported for security reasons.`,
				header.name, header.linkname,
			)}
		}

		size := header.size
		if header.typeflag == '5' {
			size = 0
		}
		if size > maxEntryBytes {
			return nil, "", &Error{Message: fmt.Sprintf("Tar entry too large: %s (%d bytes)", header.name, size)}
		}

		var content []byte
		if size > 0 {
			end := offset + int(size)
			if end > len(archive) {
				return nil, "", &Error{Message: fmt.Sprintf("Truncated tar entry: %s", header.name)}
			}
			content = append([]byte(nil), archive[offset:end]...)
		}

		entryType := types.TarEntryFile
		if header.typeflag == '5' {
			entryType = types.TarEntryDirectory
		}

		entries = append(entries, types.TarEntry{
			Path:      header.name,
			Content:   content,
			Mode:      header.mode,
			Mtime:     header.mtime * 1000,
			EntryType: entryType,
		})

		dataBlocks := int((size + blockSize - 1) / blockSize)
		offset += dataBlocks * blockSize
	}

	return entries, checksum, nil
}

// ValidateTarPath ensures a tar entry path cannot escape the extraction root.
func ValidateTarPath(entryPath, root string) (string, error) {
	if strings.Contains(entryPath, "..") {
		return "", &Error{Message: fmt.Sprintf(`Path traversal detected: "%s" contains ".."`, entryPath)}
	}
	if strings.ContainsRune(entryPath, '\x00') {
		return "", &Error{Message: fmt.Sprintf(`Path traversal detected: "%s" contains null byte`, entryPath)}
	}
	if filepath.IsAbs(entryPath) {
		return "", &Error{Message: fmt.Sprintf(`Path traversal detected: "%s" is absolute`, entryPath)}
	}

	resolved := filepath.Join(root, entryPath)
	normalizedRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		normalizedRoot = root
	}
	normalizedResolved, err := filepath.EvalSymlinks(resolved)
	if err != nil {
		normalizedResolved = resolved
	}

	if normalizedResolved != normalizedRoot &&
		!strings.HasPrefix(normalizedResolved, normalizedRoot+string(os.PathSeparator)) {
		return "", &Error{Message: fmt.Sprintf(`Path traversal detected: "%s" resolves outside root`, entryPath)}
	}

	return entryPath, nil
}

type tarHeader struct {
	name     string
	mode     uint32
	size     uint64
	mtime    uint64
	typeflag byte
	linkname string
}

func encodeHeader(entry types.TarEntry) ([]byte, error) {
	buf := make([]byte, headerSize)
	name := entry.Path

	writeField(buf, 0, 100, name)
	writeField(buf, 100, 8, fmt.Sprintf("%07o", entry.Mode&0o7777))
	writeField(buf, 108, 8, "0000000")
	writeField(buf, 116, 8, "0000000")

	sizeField := "00000000000"
	if entry.EntryType == types.TarEntryFile {
		sizeField = fmt.Sprintf("%011o", len(entry.Content))
	}
	writeField(buf, 124, 12, sizeField)
	writeField(buf, 136, 12, fmt.Sprintf("%011o", entry.Mtime/1000))
	writeField(buf, 148, 8, strings.Repeat(" ", 8))

	typeflag := byte('0')
	if entry.EntryType == types.TarEntryDirectory {
		typeflag = '5'
	}
	buf[156] = typeflag
	writeField(buf, 257, 6, magic)
	writeField(buf, 263, 2, "00")

	var checksum uint32
	for _, b := range buf {
		checksum += uint32(b)
	}
	writeField(buf, 148, 7, fmt.Sprintf("%07o", checksum))
	buf[155] = ' '

	return buf, nil
}

func decodeHeader(buf []byte) (*tarHeader, error) {
	name := readField(buf, 0, 100)
	mode := uint32(parseOctalField(readField(buf, 100, 8)))
	size := parseOctalField(readField(buf, 124, 12))
	mtime := parseOctalField(readField(buf, 136, 12))
	checksumField := readField(buf, 148, 8)
	typeflag := byte('0')
	if len(buf) > 156 && buf[156] != 0 {
		typeflag = buf[156]
	}
	linkname := readField(buf, 157, 100)

	savedChecksum := parseOctalFieldU32(strings.TrimSpace(checksumField))
	recomputed := append([]byte(nil), buf...)
	writeField(recomputed, 148, 8, strings.Repeat(" ", 8))
	var actual uint32
	for _, b := range recomputed {
		actual += uint32(b)
	}
	if savedChecksum != 0 && savedChecksum != actual {
		return nil, &Error{Message: fmt.Sprintf(
			`Checksum mismatch for tar entry "%s": expected %s, got %s`,
			name, formatOctal(actual), formatOctal(savedChecksum),
		)}
	}

	return &tarHeader{
		name:     name,
		mode:     mode,
		size:     size,
		mtime:    mtime,
		typeflag: typeflag,
		linkname: linkname,
	}, nil
}

func writeField(buf []byte, offset, length int, value string) {
	end := offset + length
	if end > len(buf) {
		return
	}
	for i := offset; i < end; i++ {
		buf[i] = 0
	}
	copy(buf[offset:end], []byte(value))
}

func readField(buf []byte, offset, length int) string {
	end := offset + length
	if end > len(buf) {
		end = len(buf)
	}
	nullPos := offset
	for nullPos < end && buf[nullPos] != 0 {
		nullPos++
	}
	return string(buf[offset:nullPos])
}

func parseOctalField(value string) uint64 {
	trimmed := strings.Trim(strings.Trim(value, "\x00"), " ")
	if trimmed == "" {
		return 0
	}
	parsed, err := strconv.ParseUint(trimmed, 8, 64)
	if err != nil {
		return 0
	}
	return parsed
}

func parseOctalFieldU32(value string) uint32 {
	return uint32(parseOctalField(value))
}

func formatOctal(value uint32) string {
	return strconv.FormatUint(uint64(value), 8)
}

func isZeroBlock(buf []byte) bool {
	for _, b := range buf {
		if b != 0 {
			return false
		}
	}
	return true
}

func hexDigest(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

File: crates/hem-core/src/tar.rs
//! Minimal POSIX ustar tar writer/reader for `.hem` bundles.

use std::fs;
use std::io;
use std::path::Path;

use sha2::{Digest, Sha256};

use crate::types::{TarEntry, TarEntryType};

const BLOCK_SIZE: usize = 512;
const HEADER_SIZE: usize = 512;
const MAGIC: &str = "ustar";
const MAX_ENTRY_BYTES: u64 = 500 * 1024 * 1024;

#[derive(Debug, thiserror::Error)]
pub enum TarError {
    #[error("{0}")]
    Message(String),
    #[error(transparent)]
    Io(#[from] io::Error),
}

pub type TarResult<T> = Result<T, TarError>;

#[derive(Debug, Clone)]
struct TarHeader {
    name: String,
    mode: u32,
    uid: u32,
    gid: u32,
    size: u64,
    mtime: u64,
    checksum: String,
    typeflag: char,
    linkname: String,
    magic: String,
    version: String,
    uname: String,
    gname: String,
    devmajor: u32,
    devminor: u32,
    prefix: String,
}

/// Write a tar archive to a file and return the SHA-256 hex digest of the archive.
pub fn write_tar(entries: &[TarEntry], output_path: &Path) -> TarResult<String> {
    let mut blocks: Vec<u8> = Vec::new();

    for entry in entries {
        let header = encode_header(entry)?;
        blocks.extend_from_slice(&header);

        if entry.entry_type == TarEntryType::File && !entry.content.is_empty() {
            blocks.extend_from_slice(&entry.content);
            let padding = BLOCK_SIZE - (entry.content.len() % BLOCK_SIZE);
            if padding < BLOCK_SIZE {
                blocks.extend(std::iter::repeat_n(0u8, padding));
            }
        }
    }

    blocks.extend(std::iter::repeat_n(0u8, BLOCK_SIZE));
    blocks.extend(std::iter::repeat_n(0u8, BLOCK_SIZE));

    if let Some(parent) = output_path.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(output_path, &blocks)?;

    Ok(hex_digest(&blocks))
}

/// Read a tar archive from a file.
pub fn read_tar(input_path: &Path) -> TarResult<(Vec<TarEntry>, String)> {
    let archive = fs::read(input_path)?;
    let checksum = hex_digest(&archive);
    let mut entries = Vec::new();
    let mut offset = 0usize;

    while offset + HEADER_SIZE <= archive.len() {
        let header_block = &archive[offset..offset + HEADER_SIZE];

        if is_zero_block(header_block) {
            offset += BLOCK_SIZE;
            if offset + BLOCK_SIZE <= archive.len()
                && is_zero_block(&archive[offset..offset + BLOCK_SIZE])
            {
                break;
            }
            continue;
        }

        let header = decode_header(header_block)?;
        offset += HEADER_SIZE;

        if header.typeflag != '0' && header.typeflag != '5' {
            return Err(TarError::Message(format!(
                "Unsupported tar entry type \"{}\" for \"{}\". \
                 Only regular files (typeflag '0') and directories (typeflag '5') are accepted.",
                header.typeflag, header.name
            )));
        }

        if !header.linkname.is_empty() {
            return Err(TarError::Message(format!(
                "Tar entry \"{}\" has a non-empty linkname field (\"{}\"). \
                 Link entries are not supported for security reasons.",
                header.name, header.linkname
            )));
        }

        let size = if header.typeflag == '5' { 0 } else { header.size };

        if size > MAX_ENTRY_BYTES {
            return Err(TarError::Message(format!(
                "Tar entry too large: {} ({} bytes)",
                header.name, size
            )));
        }

        let content = if size > 0 {
            let end = offset + size as usize;
            if end > archive.len() {
                return Err(TarError::Message(format!(
                    "Truncated tar entry: {}",
                    header.name
                )));
            }
            archive[offset..end].to_vec()
        } else {
            Vec::new()
        };

        entries.push(TarEntry {
            path: header.name,
            content,
            mode: header.mode,
            mtime: header.mtime.saturating_mul(1000),
            entry_type: if header.typeflag == '5' {
                TarEntryType::Directory
            } else {
                TarEntryType::File
            },
        });

        let data_blocks = size.div_ceil(BLOCK_SIZE as u64) as usize;
        offset += data_blocks * BLOCK_SIZE;
    }

    Ok((entries, checksum))
}

/// Validate that a tar entry path is safe (no path traversal).
pub fn validate_tar_path(entry_path: &str, root: &Path) -> TarResult<String> {
    if entry_path.contains("..") {
        return Err(TarError::Message(format!(
            "Path traversal detected: \"{entry_path}\" contains \"..\""
        )));
    }
    if entry_path.contains('\0') {
        return Err(TarError::Message(format!(
            "Path traversal detected: \"{entry_path}\" contains null byte"
        )));
    }
    if Path::new(entry_path).is_absolute() {
        return Err(TarError::Message(format!(
            "Path traversal detected: \"{entry_path}\" is absolute"
        )));
    }

    let resolved = root.join(entry_path);
    let normalized_root = root.canonicalize().unwrap_or_else(|_| root.to_path_buf());
    let normalized_resolved = resolved.canonicalize().unwrap_or(resolved);

    let root_str = normalized_root.to_string_lossy();
    let resolved_str = normalized_resolved.to_string_lossy();
    if resolved_str != root_str && !resolved_str.starts_with(&format!("{root_str}{}", MAIN_SEPARATOR)) {
        return Err(TarError::Message(format!(
            "Path traversal detected: \"{entry_path}\" resolves outside root"
        )));
    }

    Ok(entry_path.to_string())
}

fn encode_header(entry: &TarEntry) -> TarResult<Vec<u8>> {
    let mut buf = vec![0u8; HEADER_SIZE];
    let name = &entry.path;
    let prefix = "";

    write_field(&mut buf, 0, 100, name);
    write_field(&mut buf, 100, 8, &format!("{:07o}", entry.mode & 0o7777));
    write_field(&mut buf, 108, 8, "0000000");
    write_field(&mut buf, 116, 8, "0000000");
    let size_field = if entry.entry_type == TarEntryType::File {
        format!("{:011o}", entry.content.len())
    } else {
        "00000000000".to_string()
    };
    write_field(&mut buf, 124, 12, &size_field);
    write_field(
        &mut buf,
        136,
        12,
        &format!("{:011o}", entry.mtime / 1000),
    );
    write_field(&mut buf, 148, 8, &" ".repeat(8));
    let typeflag = if entry.entry_type == TarEntryType::Directory {
        '5'
    } else {
        '0'
    };
    write_field(&mut buf, 156, 1, &typeflag.to_string());
    write_field(&mut buf, 157, 100, "");
    write_field(&mut buf, 257, 6, MAGIC);
    write_field(&mut buf, 263, 2, "00");
    write_field(&mut buf, 265, 32, "");
    write_field(&mut buf, 297, 32, "");
    write_field(&mut buf, 329, 8, "0000000");
    write_field(&mut buf, 337, 8, "0000000");
    write_field(&mut buf, 345, 155, prefix);

    let checksum: u32 = buf.iter().map(|&b| u32::from(b)).sum();
    let checksum_str = format!("{:07o}", checksum);
    write_field(&mut buf, 148, 7, &checksum_str);
    buf[155] = 0x20;

    Ok(buf)
}

fn decode_header(buf: &[u8]) -> TarResult<TarHeader> {
    let name = read_field(buf, 0, 100);
    let mode = parse_octal_field(read_field(buf, 100, 8));
    let uid = parse_octal_field(read_field(buf, 108, 8));
    let gid = parse_octal_field(read_field(buf, 116, 8));
    let size = parse_octal_field(read_field(buf, 124, 12));
    let mtime = parse_octal_field(read_field(buf, 136, 12));
    let checksum = read_field(buf, 148, 8);
    let typeflag = read_field(buf, 156, 1)
        .chars()
        .next()
        .unwrap_or('\0');
    let linkname = read_field(buf, 157, 100);
    let magic = read_field(buf, 257, 6);
    let version = read_field(buf, 263, 2);
    let uname = read_field(buf, 265, 32);
    let gname = read_field(buf, 297, 32);
    let devmajor = parse_octal_field(read_field(buf, 329, 8));
    let devminor = parse_octal_field(read_field(buf, 337, 8));
    let prefix = read_field(buf, 345, 155);

    let saved_checksum = parse_octal_field_u32(checksum.trim());
    let computed: u32 = buf.iter().map(|&b| u32::from(b)).sum();
    let mut recomputed_buf = buf.to_vec();
    write_field(&mut recomputed_buf, 148, 8, &" ".repeat(8));
    let actual: u32 = recomputed_buf.iter().map(|&b| u32::from(b)).sum();

    if saved_checksum != 0 && saved_checksum != actual {
        if saved_checksum != computed && saved_checksum != actual {
            return Err(TarError::Message(format!(
                "Checksum mismatch for tar entry \"{name}\": expected {}, got {}",
                format_octal(actual),
                format_octal(saved_checksum)
            )));
        }
    }

    Ok(TarHeader {
        name,
        mode: mode as u32,
        uid: uid as u32,
        gid: gid as u32,
        size,
        mtime,
        checksum,
        typeflag,
        linkname,
        magic,
        version,
        uname,
        gname,
        devmajor: devmajor as u32,
        devminor: devminor as u32,
        prefix,
    })
}

fn write_field(buf: &mut [u8], offset: usize, length: usize, value: &str) {
    let end = offset + length;
    let slice = &mut buf[offset..end];
    slice.fill(0);
    let bytes = value.as_bytes();
    let copy_len = bytes.len().min(length);
    slice[..copy_len].copy_from_slice(&bytes[..copy_len]);
}

fn read_field(buf: &[u8], offset: usize, length: usize) -> String {
    let end = offset + length;
    let mut null_pos = offset;
    while null_pos < end && buf[null_pos] != 0 {
        null_pos += 1;
    }
    String::from_utf8_lossy(&buf[offset..null_pos]).into_owned()
}

fn parse_octal_field(value: String) -> u64 {
    let trimmed = value.trim_matches('\0').trim();
    if trimmed.is_empty() {
        0
    } else {
        u64::from_str_radix(trimmed, 8).unwrap_or(0)
    }
}

fn parse_octal_field_u32(value: &str) -> u32 {
    let trimmed = value.trim_matches('\0').trim();
    if trimmed.is_empty() {
        0
    } else {
        u32::from_str_radix(trimmed, 8).unwrap_or(0)
    }
}

fn format_octal(value: u32) -> String {
    format!("{value:o}")
}

fn is_zero_block(buf: &[u8]) -> bool {
    buf.iter().all(|&b| b == 0)
}

fn hex_digest(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hasher
        .finalize()
        .iter()
        .map(|byte| format!("{byte:02x}"))
        .collect()
}

#[cfg(unix)]
const MAIN_SEPARATOR: char = '/';

#[cfg(windows)]
const MAIN_SEPARATOR: char = '\\';